crypto/o_str.c

```diff
@@ -223,7 +223,22 @@ int openssl_strerror_r(int errnum, char *buf, size_t buflen)
 #if defined(_MSC_VER) && _MSC_VER>=1400
     return !strerror_s(buf, buflen, errnum);
 #elif defined(_GNU_SOURCE)
-    return strerror_r(errnum, buf, buflen) != NULL;
+    char *err;
+
+    /*
+     * GNU strerror_r may not actually set buf.
+     * It can return a pointer to some (immutable) static string in which case
+     * buf is left unused.
+     */
+    *buf = '\0';
+    err = strerror_r(errnum, buf, buflen);
+    if (err == NULL)
+        return 0;
+    if (*buf == '\0') {
```

Comment:
did you consider `if (err != buf)` as alternative condition here ?

Reply:
I made the same comment, but it seems there are no guarantees given in the man page that err == buf in the event the buffer is filled (I can't think why they would do this, but it may not be aligned with the start of the buffer).

Reply:
So if err is somewhere within the buffer, then OPENSSL_strlcpy should be safe,
while strncpy is not guaranteed to work on overlapped in and out.

Reply:
It wasn't clear to me that `err == buf` from the system man page. However, I've found this in the [glibc documentation](https://www.gnu.org/software/libc/manual/html_node/Error-Messages.html):

`This (err) might be either some permanent global data or a message string in the user supplied buffer starting at buf with the length of n bytes.`

That would make the `if (err != buf)` condition viable.

Reply:
It depends how you read the line. You could read it as the user supplied buffer starts at buf and has length n. Not necessarily that err *starts* at buf.

Reply:
You're probably right. I kept the current condition and changed the code to use 
OPENSSL_strlcpy.

Reply:
.. and change this to "if (err != buf) {"

Reply:
@bernd-edlinger - I think this is incorrect for the reasons I give above

Reply:
of course the braces can go away.

Reply:
@mattcaswell my reasoning is if err is statically allocated, it is != buf, and copying it to buf will do the right thing.
if err points somewhere inside buf, it OPENSL_strlcpy can handle this, since src, and dest are not annotated
with __restrict, and the function reads src byte for byte, and writes to dest. so that will work too.
if err == buf  we do not have to copy anything, so that will also be correct.
It might be possible to add a comment of course.

Reply:
ah - ok. That makes sense. A comment explaining that would be a really good idea!

Reply:
The code has been updated, including the comment.